students/TrkCat/Lesson06/test_mailroom.py

```diff
@@ -0,0 +1,148 @@
+                                       'last_don': 1,
+                                       'name': 'Abe Lincoln',
+                                       'num_don': 3,
+                                       'tot_don': 8
+                                       }
+                       }
+    mailroom.add_donation('Abe Lincoln', 8)
+    assert mailroom.donors['abe lincoln']['last_don'] == 8
+    assert mailroom.donors['abe lincoln']['num_don'] == 4
+
+
+def test2_add_donation():
+    mailroom.donors = {'abe lincoln': {'avg_don': 2.6666666666666665,
+                                       'donations': [5, 2, 1],
+                                       'last_don': 1,
+                                       'name': 'Abe Lincoln',
+                                       'num_don': 3,
+                                       'tot_don': 8
+                                       }
+                       }
+    mailroom.add_donation('Testy McTestFace', 50000)
+    assert 'testy mctestface' in mailroom.donors
+    assert mailroom.donors['testy mctestface']['name'] == 'Testy McTestFace'
+    assert mailroom.donors['testy mctestface']['last_don'] == 50000
+    assert mailroom.donors['testy mctestface']['num_don'] == 1
+    assert mailroom.donors['testy mctestface']['tot_don'] == 50000
+    assert mailroom.donors['testy mctestface']['avg_don'] == 50000
+
+
+def test_write_letters_to_file():
+    mailroom.donors = {'abe lincoln': {'avg_don': 2.6666666666666665,
+                                       'donations': [5, 2, 1],
+                                       'last_don': 1,
+                                       'name': 'Abe Lincoln',
+                                       'num_don': 3,
+                                       'tot_don': 8
+                                       },
+                       'yankee doodle': {'avg_don': 67.0,
+                                         'donations': [67],
+                                         'last_don': 67,
+                                         'name': 'Yankee Doodle',
+                                         'num_don': 1,
+                                         'tot_don': 67
+                                         }
+                       }
+    mailroom.write_letters_to_file('.')
+    assert os.path.isfile('Abe Lincoln.txt')
+    assert os.path.isfile('Yankee Doodle.txt')
+
+    with open('Abe Lincoln.txt', 'r') as f:
+        actual = f.read()
+
+    expected = ('FROM: Your friendly local charity mailroom.\n'
+                'TO: Abe Lincoln\n'
+                'RE: Your recent donation\n'
+                '\n'
+                '\n'
+                'Thank you so much for your recent donation of $1.00. This '
+                'will go a long way towards helping to save the pythons. '
+                'Your generosity is most appreciated!\n'
+                '\n'
+                'Best Regards,\n'
+                'Save The Pythons\n'
+                )
+    assert expected == actual
```

Comment:
I know you mentioned that you struggled with this exercise but this is really excellent work.

You have a good variety of tests that are of good quality and something that I would have expected in a real production code. Nice work!